### PR TITLE
Proof of concept: typescript transformer support for tsickle

### PR DESCRIPTION
**Pull request by tbosch**
Needs:

- [x] review of general direction
- [x] test in G3
- [x] add tests for source maps with transformers
- [x] split into multiple meaningful commits for the review

### CODE REVIEW

**Comment by mprobst** [comment]
Thanks for the change! This is very interesting.

I'm not quite sure how to make progress on this though – as is, this is just a huge lump of code (and overall in not quite reviewable state, AFAICT). Could you maybe write up a doc explaining the basic approach, and giving a guided tour through the changes? We could also have a meeting and walk through the code.

**Comment by orenmizr** [comment]
Hi, i'm just watching progress on this project. what is change (typescript transformer) about ? #curious

**Comment by evmar** [comment]
https://github.com/Microsoft/TypeScript/issues/5595 is the main idea I
think, this random blog post goes into it
http://blog.scottlogic.com/2017/05/02/typescript-compiler-api-revisited.html

File: src/rewriter.ts [hunk 1]
```diff
@@ -71,58 +68,59 @@ export abstract class Rewriter {
...
-      // section preceding it. That's arguably incorrect (e.g. for the fake
-      // methods defining properties), but is good enough for stack traces.
-      const pos = this.file.getLineAndCharacterOfPosition(from);
-      this.sourceMap.addMapping({
-        original: {line: pos.line + 1, column: pos.character + 1},
-        generated: this.position,
-        source: this.file.fileName,
-      });
+      if (text.indexOf('A Used by implement_import.ts') !== -1) {
+        console.log('>>> now', node.kind, new Error().stack);
```
Comment (tbosch) [inline src/rewriter.ts]: remove this

**Comment by tbosch** [comment]
Prefactor PR is here: https://github.com/angular/tsickle/pull/514

**Comment by tbosch** [comment]
Final PR is here: https://github.com/angular/tsickle/pull/516

**Pull request closed without merging**